Significant API rework and no_std support

Amanieu opened the pull request:
This is a WIP PR which does many things:

- Makes the `SkipList` API usable with `no_std` by requiring a `Guard` to be passed in where needed (#4).
- Adds `lower_bound` and `upper_bound` functions (#5).
- Significant performance improvements for lookups (3x faster on ARM).


Since a lot of changes were made to the code, I put the new code in `base1.rs`.

Amanieu committed:
Significant API rework and no_std support

Fixes #4
Fixes #5

stjepang commented on src/base1.rs:
```diff
@@ -0,0 +1,1947 @@
+}
+
+/// An entry in a skip list, protected by a `Guard`.
+pub struct Entry<'a, 'g, K: 'a + 'g, V: 'a + 'g> {
```
Is there any point in having two lifetimes? :) Why not define `front` and other methods like this:

```rust
fn front<'a>(&'a self, guard: &'a Guard) -> Option<Entry<'a, K, V>>
```

Amanieu replied:
The guard and self lifetimes needs to be separate for `pin` to work.

stjepang commented on src/base1.rs:
```diff
@@ -0,0 +1,1947 @@
+    ///
+    /// This method may return `None` if the reference count is already 0 and
+    /// the node has been queued for deletion.
+    pub fn pin(&self) -> Option<RefEntry<'a, K, V>> {
```
I'd prefer a different name for this method. Not sure which one, though...

stjepang commented on src/base1.rs:
```diff
@@ -0,0 +1,1947 @@
+///
+/// You *must* call `release` to free this type, otherwise the node will be
+/// leaked. This is because releasing the entry requires a `Guard`.
+pub struct RefEntry<'a, K: 'a, V: 'a> {
```
If we implement "anonymous" pinning, then we'll be able to automatically release when dropping `RefEntry`:

Amanieu replied:
Are there any concrete plans for anonymous pinning? How would this work in practice?

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1948 @@
+
+    /// Hot data associated with the skip list, stored in a dedicated cache line.
+    hot_data: CachePadded<HotData>,
+}
```
Since the head node is now embedded directly in `SkipList`, the actual size is even larger: `sizeof(usize) * MAX_HEIGHT = 256`.

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1947 @@
+    where
+        K: Borrow<Q>,
+        Q: Ord + ?Sized,
+    {
```
`search_bound` is meant to be the "fast" version which is used for lookups, while `search_position` is meant to be the "slow" version used for insert/delete.

stjepang replied:
Is there really any difference in performance? I'd imagine the compiler can optimize away any apparent differences.

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1947 @@
+                        Node::finalize(node.as_raw());
+                    });
+                    search = self.search_position(&n.key, guard);
+                    mem::forget(sg);
```
`defer_on_unwind` isn't available in `no_std`. It's also somewhat slow since it checks thread-local state to see if we are currently unwinding.

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1948 @@
+pub struct Iter<'a, 'g, K: 'a + 'g, V: 'a + 'g> {
+    parent: &'a SkipList<K, V>,
+    node: Option<&'g Node<K, V>>,
+    guard: &'g Guard,
```
It avoids the need to mess with the `Guard` reference count. But I don't think that would have much impact on performance.

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1948 @@
+        lower_bound: Bound<&'k Min>,
+        upper_bound: Bound<&'k Max>,
+        guard: &'g Guard,
+    ) -> Range<'a, 'g, 'k, Min, Max, K, V>
```
I actually modeled this off the old `range` API: https://doc.rust-lang.org/1.3.0/collections/struct.BTreeMap.html#method.range

I personally like it better than the new API.

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1948 @@
+        &self,
+        pred: &Atomic<Node<K, V>>,
+        curr: &'g Node<K, V>,
+        succ: Shared<'g, Node<K, V>>,
```
I agree that `curr` doesn't need to be bound by `'g`, however `succ` definitely does since that is what the return value is based on.

Amanieu commented on src/base1.rs:
```diff
@@ -0,0 +1,1947 @@
+        // Rebind the guard to the lifetime of self. This is a bit of a
+        // hack but it allows us to return references that are not bound to
+        // the lifetime of the guard.
+        let guard = &*(guard as *const _);
```
Yea, the logic here is that `search_position` is used by `insert` and `remove`, which both return a `RefEntry` instead of an `Entry`.

Amanieu commented:
All comments have been addressed.

Amanieu commented:
For `SeqCst`, I think this should be done by adding `SeqCst` fences around skip list operations rather than by changing the internal orderings.

Amanieu commented:
Here are some benchmark results. Each test performs 1000 operations on the map.

### x86
SkipList
```
test insert        ... bench:     195,518 ns/iter (+/- 13,275)
test insert_remove ... bench:     425,977 ns/iter (+/- 30,948)
test iter          ... bench:       4,183 ns/iter (+/- 284)
test lookup        ... bench:      76,276 ns/iter (+/- 4,240)
```
BTreeMap
```
test insert        ... bench:      94,297 ns/iter (+/- 7,352)
test insert_remove ... bench:     197,884 ns/iter (+/- 58,158)
test iter          ... bench:       3,768 ns/iter (+/- 304)
test lookup        ... bench:      72,858 ns/iter (+/- 2,003)
```
HashMap
```
test insert        ... bench:      75,179 ns/iter (+/- 4,698)
test insert_remove ... bench:     102,511 ns/iter (+/- 5,446)
test iter          ... bench:       1,424 ns/iter (+/- 161)
test lookup        ... bench:      19,948 ns/iter (+/- 1,076)
```
### AArch64
SkipList
```
test insert        ... bench:     583,522 ns/iter (+/- 1,488)
test insert_remove ... bench:   1,167,879 ns/iter (+/- 5,897)
test iter          ... bench:      10,756 ns/iter (+/- 8)
test lookup        ... bench:     176,297 ns/iter (+/- 326)
```
BTreeMap
```
test insert        ... bench:     208,406 ns/iter (+/- 267)
test insert_remove ... bench:     384,687 ns/iter (+/- 472)
test iter          ... bench:       8,740 ns/iter (+/- 36)
test lookup        ... bench:     117,029 ns/iter (+/- 223)
```
HashMap
```
test insert        ... bench:     149,866 ns/iter (+/- 2,169)
test insert_remove ... bench:     209,747 ns/iter (+/- 2,152)
test iter          ... bench:       8,963 ns/iter (+/- 18)
test lookup        ... bench:      51,145 ns/iter (+/- 39)
```

stjepang commented on src/base1.rs:
```diff
@@ -0,0 +1,1985 @@
+    /// If the skip list is being concurrently modified, consider the returned number just an
+    /// approximation without any guarantees.
+    pub fn len(&self) -> usize {
+        self.hot_data.len.load(Ordering::Relaxed)
```
Since `len` is just an approximation updated with `Relaxed` ordering and could sometimes be a little off, I fear concurrent inserts and deletes might sometimes make the number fall below zero, which would make it a huge number due to underflow.

The fix is easy:

```rust
let len = self.hot_data.len.load(Ordering::Relaxed);
if len > isize::MAX as usize {
    0
} else {
    len
}
```

stjepang commented on src/base1.rs:
```diff
@@ -0,0 +1,1985 @@
+/// An entry in a skip list, protected by a `Guard`.
+///
+/// The lifetimes of the key and value are the same as that of the `Guard`
+/// used when creating the `Entry` (`'g`). This lifetime is also contrained to
```
s/contrained/constrained

stjepang commented:
> - Rename `Entry::{next, get_next}` to `Entry::{move_next, next}`

Both options look fine to me. 👍

> - Remove `SkipMap::pop_back`
> - Remove `Entry::{prev, get_prev}`
> - Remove `DoubleEndedIterator` from `Iter`
>
> The main change is that I am removing backward iteration since it represents a hidden `O(log(n))` cost on every step. This means that iterating backwards is `O(n log(n))`, which may not be obvious to a user.

That sounds like a harsh price to pay. I think simply stating in the documentation that going forwards is fast and going backwards is slow will be enough. That's just a fundamental property of `crossbeam-skiplist`.

LevelDB similarly allows iterating forwards and backwards and mentions the caveat in its [documentation](https://github.com/google/leveldb/blob/master/doc/index.md#iteration).

Amanieu commented:
OK, I have everything implemented except `SkipSet`. Since there's a quite a few changes, a look-over would be nice. In particular:
- I ported `SkipMap` to use the new `SkipList`. New methods on `SkipMap`: `lower_bound` (replaces `seek`), `upper_bound`, `range`.
- I added `base::RefIter` and `base::RefRange` which returned `RefEntry` instead of `Entry`. These are used by the `SkipMap` iterators.
- All iterator types (`Iter`, `RefIter`, `Range`, `RefRange`) except `IntoIter` are now bi-directional.
- I have added back methods for backwards iterator (`SkipList::pop_back`, `Entry::prev`, etc).

Amanieu commented:
And `SkipSet` is done now. Everything is ready for merging! (finally)

Amanieu commented:
> Quick question: Are we also going to add fences to `SkipMap` and `SkipSet` for linearizability?

I would prefer leaving it as it is. I still feel that linearizability is not worth the added cost & complexity.

stjepang commented:
> I would prefer leaving it as it is. I still feel that linearizability is not worth the added cost & complexity.

If the skip list is linearizable, that is arguably less complexity for the user to think about. And for us, the implementors, the lack of linearizability in `base.rs` means we have to be extra careful and ask ourselves twice whether it is safe to use anything other than `SeqCst` every time we do. It's sometimes a risky gamble, but hopefully we got everything right. That's more complexity than if we just went with `SeqCst` everywhere.

But if by complexity you meant the added fences to the `SkipMap` and `SkipSet` methods before/after forwarding calls to `base.rs`, then it should be relatively simple. Or am I missing something?

It is true that adding an option for linearizability increases the amount of code and maybe the API surface, however. And yes, having both relaxed and linearizable versions of the skip list is more complex than having just one version. We're on the same page on that one.

Furthermore, if we don't make `SkipMap` and `SkipSet` linearizable, then how does one get linearizability themselves? In addition to linearizability not being the default, now it's also difficult/error-prone to achieve at all. This is a non-starter for me.

My disagreement on this issue is twofold:

1. Linearizability should not be a luxury, it should at the very least be **an option**.
2. Moreover, it should not be just an extra option, it should be **the default**.

I think this philosophy (correctness by default, performance by request) has countless examples in Rust and in other languages:

* `ConcurrentSkipListMap` in Java is fully seqcst. There isn't even an option for more relaxed operations.
* `slice::sort` is the stable sort. The faster and unstable sort is `sort_unstable`.
* `slice::get` is the safe access method. The faster and unsafe version is `slice::get_unchecked`.
* `compare_and_set` and `compare_exchange` are strong variants. The faster and weaker one is `compare_exchange_weak`.
* In C++, all operations on `std::atomic` are seqcst by default. If Rust had default arguments, they would be, too.
* It is impossible to break the safety of safe Rust. If you need to go faster, just write `unsafe { ... }`.
* You can get shared mutability with `Cell`, `RefCell`, `Mutex`, `RwLock`. Too slow? - No problem, use `unsafe`.
* Other synchronization primitives like `Mutex` and channels are sequentially consistent, too.

So the philosophy is always that safety/consistency/correctness is the default, and performance is an option. I'd argue that the great advance of Rust is that it offers performance despite safety, not safety despite performance. And if we ever really have to choose just one between the two, safety trumps performance. (Of course, there's some nuance around things like arithmetic overflow in release mode, but those are very minor deviations, and don't really break safety.)

Note that this sometimes even applies to C++, an unsafe language which usually goes the other way around and prefers to sacrifice safety over performance. In C++, seqcst operations in `std::atomic` are the default. Also, it chooses to offer only the equivalent of `Arc`, and `Rc` is not even an option!

Going back to our skip lists, I'd say we have to have sequentially consistent skiplists, one way or another. This is not to say that faster and more relaxed operations shouldn't be offered - just not as the only nor the default option. Relaxed operations could be exposed through `base::SkipList`, or `weak::SkipMap`, or `SkipMap::<operation>_weak`, or something like that.

I think we've had similar discussions before. Do you think that despite the precedent stemming from the examples I've listed, we still shouldn't have linearizable skiplist-based maps and sets? If so, why? (Yes, there's a performance cost, but we're still being faster than Java!).

Finally, here's my go-to example illustrating what goes wrong when you omit linearizability. Consider the following:

```rust
let q = ConcurrentQueue::new(2); // the capacity is 2
crossbeam::scope(|s| {
    for _ in 0..2 {
        s.spawn(|| {
            for _ in 0..10_000 {
                q.push(());
                assert_eq!(q.try_pop(), Some(()));
            }
        });
    }
});
```

The assert will break if you implement [this](http://www.1024cores.net/home/lock-free-algorithms/queues/bounded-mpmc-queue) concurrent queue (one of the most popular!). This could lead to really subtle bugs and the real problem is difficult to understand, let alone foresee. And that's just a really simple queue...

Amanieu commented:
In the end, I'm only going to be using `base::SkipList` myself, so I don't really care. If you feel strongly about it then let's just add this as an option.

Regarding how we can implement this option, here is what I propose:
```rust
trait Linearizable {
    fn seq_cst_fence();
}
struct Relaxed;
impl Linearizable for Relaxed {
    fn seq_cst_fence() {}
}
struct SeqCst;
impl Linearizable for SeqCst {
    fn seq_cst_fence() { fence(SeqCst); }
}

struct SkipMap<K, V, L: Linearizable = SeqCst> { ... }
```

stjepang commented:
Ok, so I want to give this PR a thumbs up now, and revisit the linearizability issue later. Since we're in disagreement, I wonder what the general public thinks about it. But let's figure this out later.

Feel free to merge!

Amanieu commented:
Can you open an issue to track the linearizability issue?
